src/core/inventory.ts

```diff
@@ -0,0 +1,371 @@
+  let balance: bigint
+  let decimals: number
+
+  // TODO: Not every chain's native token is 0 address...
+  // Handle native token (ETH, MATIC, etc.)
+  if (tokenAddress === '0x0000000000000000000000000000000000000000') {
+    try {
+      balance = await publicClient.getBalance({ address: account })
+      // Native tokens typically have 18 decimals
+      decimals = 18
+    } catch (error) {
+      console.error(
+        `Error getting native token balance on chain ${chainId}:`,
+        error,
+      )
+      balance = 0n
+      decimals = 18
+    }
+  } else {
+    // For ERC-20 tokens
+    try {
+      // Get balance
+      const balanceResult = await publicClient.readContract({
+        address: tokenAddress,
+        abi: erc20ABI,
+        functionName: 'balanceOf',
+        args: [account],
+      })
+
+      balance = balanceResult as bigint
+
+      // Try to get decimals from the contract
+      try {
+        const decimalsResult = await publicClient.readContract({
+          address: tokenAddress,
+          abi: erc20ABI,
+          functionName: 'decimals',
+          args: [],
+        })
+
+        decimals = Number(decimalsResult)
+      } catch (error) {
+        // If decimals function fails, try to find in our mapping first, then fall back to default
+        console.warn(
+          `Failed to get decimals for token ${tokenAddress} on chain ${chainId}, using configured value or default`,
+        )
+
+        // Find by token address (in case it's mapped) or use symbol mapping through assetConfig
+        // If all else fails, default to 18 decimals
+        decimals = assetDecimals[tokenAddress] || 18
+      }
+    } catch (error) {
+      console.error(
+        `Error getting balance for token ${tokenAddress} on chain ${chainId}:`,
+        error,
+      )
+      balance = 0n
+      decimals = 18
+    }
+  }
+
+  return { balance, decimals }
+}
+
+/**
+ * Get balances for the relayer across all configured chains
+ * Can optionally filter to a specific asset symbol
+ */
+export async function getBalances(
+  getRPCUrl: (chainId: number) => string,
+  relayerAddress: Address,
+  assetSymbol?: string,
+): Promise<PortfolioBalance> {
+  const portfolio: PortfolioBalance = {
+    chains: [],
+    totalValueUSD: 0,
+  }
+
+  // Process each chain from our configuration
+  for (const [chainIdStr, chainConfig] of Object.entries(chains)) {
+    const chainId = parseInt(chainIdStr)
+
+    if (!chainConfig.assets) continue
+
+    const chainBalance: ChainBalance = {
+      chainId,
+      assets: [],
+      totalValueUSD: 0,
+    }
+
+    // Process each asset on this chain
+    for (const [symbol, assetConfig] of Object.entries(chainConfig.assets)) {
+      // Skip if we're filtering by asset and this isn't the one we want
+      if (assetSymbol && symbol !== assetSymbol) continue
+
+      try {
+        // Query the on-chain balance
+        const { balance, decimals } = await getTokenBalance(
```

Comment:
could we batch differrent fetchs into a single multicall per chain?

Reply:
I'm actually a bit confused here, because it seems like the DepositEvent just has one outputToken. As a result, we only need to check balance for this one token on each chain. However, if you don't specify the token you want to check balance for in getTokenBalance, it will default to checking balance for every supported asset on every chain, in which case *yes*, it should use a multicall, so we may want to insert that functionality into getTokenBalance utility at some point. Will leave a TODO

Reply:
so all the output tokens across different segments get filled in the same fill tx so can be multi token